IdentityCore/src/throttling/cache/MSIDLRUCache.m

```diff
@@ -0,0 +1,432 @@
+}
+
+- (void)addToFront:(MSIDLRUCacheNode *)node
+{
+    dispatch_barrier_sync(self.synchronizationQueue, ^{
+        [self addToFrontImpl:node];
+    });
+}
+
+- (void)addToFrontImpl:(MSIDLRUCacheNode *)node
+{
+    NSString *currentHeadSignature = self.head.nextSignature; //node currently pointed by the head
+    MSIDLRUCacheNode *currentHeadNode = [self.container objectForKey:currentHeadSignature];
+    
+    /**
+    BEFORE:
+     A->B->C
+     A<-B<-C
+    AFTER:
+     A->C
+     A<-C
+     */
+    currentHeadNode.prevSignature = [node.signature mutableCopy];
+    node.prevSignature = [HEAD_SIGNATURE mutableCopy];
+    node.nextSignature = [currentHeadSignature mutableCopy];
+    self.head.nextSignature = [node.signature mutableCopy];
+    
+    [self.container setObject:currentHeadNode forKey:currentHeadSignature];
+    [self.container setObject:node forKey:node.signature];
+}
+
+- (MSIDLRUCacheNode *)getHeadNode
+{
+    __block MSIDLRUCacheNode *node;
+    dispatch_sync(self.synchronizationQueue, ^{
+        if (![self.head.nextSignature isEqualToString:TAIL_SIGNATURE])
+        {
+            node = [self.container objectForKey:self.head.nextSignature];
+        }
+    });
+    return node;
+    
+}
+
+- (MSIDLRUCacheNode *)getTailNode
+{
+    __block MSIDLRUCacheNode *node;
+    dispatch_sync(self.synchronizationQueue, ^{
+        if (![self.tail.prevSignature isEqualToString:HEAD_SIGNATURE])
+        {
+            node = [self.container objectForKey:self.tail.prevSignature];
+        }
+    });
+    return node;
+}
+
+- (NSArray *)enumerateAndReturnAllObjects
+{
+    __block NSMutableArray *res;
+    dispatch_sync(self.synchronizationQueue, ^{
+        res = [self enumerateAndReturnAllObjectsImpl];
+    });
+    return res;
+}
+
+- (NSMutableArray *)enumerateAndReturnAllObjectsImpl
+{
+    NSMutableArray *res;
+    if (![self.head.nextSignature isEqualToString:TAIL_SIGNATURE])
+    {
+        res = [NSMutableArray new];
+        NSMutableString *signature = [self.head.nextSignature mutableCopy];
+        
+        while (![signature isEqualToString:TAIL_SIGNATURE])
+        {
+            MSIDLRUCacheNode *node = [self.container objectForKey:signature];
+            [signature setString:node.nextSignature];
+            [res addObject:node.cacheRecord];
+        }
+    }
+    return res;
+}
+
+- (NSMutableArray *)enumerateAndReturnAllNodesImpl
+{
+    NSMutableArray *res;
+    if (![self.head.nextSignature isEqualToString:TAIL_SIGNATURE])
+    {
+        res = [NSMutableArray new];
+        NSMutableString *signature = [self.head.nextSignature mutableCopy];
+        
+        while (![signature isEqualToString:TAIL_SIGNATURE])
+        {
+            MSIDLRUCacheNode *node = [self.container objectForKey:signature];
+            [signature setString:node.nextSignature];
+            [res addObject:node];
+        }
+    }
+    return res;
+}
+
+- (NSString *)generateRandomSignature //mock pointer 62^8 ~ 2^14 randomness
+{
```

Comment:
2^14 = 16384. Is it correct that: if this function generates duplicate numbers, the linked list you constructed in this class will become a graph? (any reason not to use NSUUID?)

Reply:
sorry there's a typo in the comment. this is actually 2*10^14. sorry for the confusion. 
NSUUID also works :) just wanted to have more control over the length and randomness of the pseudo-pointer.

Reply:
+1 to use NSUUID here. We can simplify the logic here and reduce complexity.